docs/source/api/top-level-await.md

```diff
@@ -0,0 +1,120 @@
+}
+```
+
+In previous versions of Meteor, async code using fibers could be run in the top level of a module. Top level await allows writing similar code that works without fibers. There are a few differences that this article will cover.
+
+Meteor's implementation of top level await tries to closely follow the specification. There currently are some differences with how Meteor handles circular dependencies.
+
+## Using top level await
+
+Top level await can be used in any app or package that uses the `ecmascript`, `typescript`, or `coffeescript` packages, or that uses any other build plugin that compiles top level await using reify.
+Generally, if you can use ECMAScript modules, then you can also use top level await.
+
+There are some extra considerations when using top level await in packages. They are covered later in this article.
+
+Top level await is only enabled by default on the server. You can enable it for the client by setting the env var `METEOR_ENABLE_CLIENT_TOP_LEVEL_AWAIT` to `true`. There are a couple known issues with using TLA on the client:
+
+1. It breaks any files in `/client/compatibility` since it now wraps those files in a function
+2. Hot module replacement has not been updated to work with TLA
+
+### What is an async module?
+
+With top level await, some modules are considered async, which affects how it behaves. There are two ways a module can become an async module:
+1. It uses top level await
+2. It imports a module that is async
+
+For example, this module (`setup.js`) would be async because it uses top level await:
+
+```js
+await setupLanguages();
+```
+
+This module (`main.js`) would be sync:
+
+```js
+console.log('in main.js');
+```
+
+However, if it imports `setup.js` which does use top level await, then `main.js` also becomes async.
+
+```js
+import './setup.js';
+
+console.log('in main.js');
+```
+
+### `require`
+
+When using `require` to load an async module, instead of directly returning a module's exports, it will return a promise that resolves to the module's exports.
+
+```js
+// resolves to the exports of init.js
+const promise = require('./init.js');
+```
+
+If you are using `require`, this does mean you need to be somewhat careful when adding/removing top level await in a file since you also have to update where the module is required.
+Since a module becomes async if it depends on an async module, this could affect more than just the individual modules using top level await.
+
+When possible, you can use ecmascript import syntax or dynamic imports instead so you don't have to worry about which modules are sync or async.
+
+### Nested imports
+
+For nested imports, it refers to using `import ...` outside of the root of a module, for example in an if block or a function.
+
+```js
+if (Meteor.isClient) {
+  import './init-client.js';
+}
+
+export function showNotification(message) {
+  import show from './notifications.js';
+
+  show(message);
+}
+```
```

Comment:
At compile, could we turn these nested imports into required imports?

Or is it already?

```js
if (Meteor.isClient) {
 require('./init-client.js');
}
export function showNotification(message) {
  const show = require('./notifications.js');
  show(message);
}
```
And if we could even, by a flag, use them like this, we could use something like SWC, I think.

The problem is that if the user relies on an async module, that is where things would go wrong. But it already throws an error.

Reply:
That is similar to what reify does, with some differences to handle live bindings.

I think we will want to move away from nested imports. The rest of the javascript ecosystem is able to work without them, and it is becoming increasingly common for new developer tools to not support them.

We could look into doing something like that with SWC, but it might not be much faster than transpiling once with just babel. What's nice about Meteor is every package and app can independently choose what build plugins to use. If some packages or apps don't use nested imports, then they could replace `ecmascript` with a build plugin that uses `swc`, while rest of the packages used by the app continue using `ecmascript`.

Reply:
Yeah, that was most of my thoughts. Many new packages, and for most apps, probably will be better with using a plugin like SWC, as they might not even know that nested import is a thing.
That does sound good. Will think more about this matter, that may be a thing to be tackled when meteor 3.x is released.